cpp/src/c_api/uniform_neighbor_sampling.cpp

```diff
@@ -222,6 +222,80 @@ extern "C" cugraph_type_erased_host_array_view_t* cugraph_sample_result_get_coun
   return reinterpret_cast<cugraph_type_erased_host_array_view_t*>(internal_pointer->count_->view());
 }
 
+extern "C" cugraph_error_code_t cugraph_sample_result_create(
+  const cugraph_resource_handle_t* handle,
+  const cugraph_type_erased_host_array_view_t* srcs,
+  const cugraph_type_erased_host_array_view_t* dsts,
+  const cugraph_type_erased_host_array_view_t* weights,
+  const cugraph_type_erased_host_array_view_t* counts,
+  cugraph_sample_result_t** result,
+  cugraph_error_t** error)
+{
+  *result = nullptr;
+  *error  = nullptr;
+  size_t n_bytes{0};
+
+  try {
+    if (!handle) {
+      *error = reinterpret_cast<cugraph_error_t*>(
+        new cugraph::c_api::cugraph_error_t{"invalid resource handle"});
+      return CUGRAPH_INVALID_HANDLE;
+    }
+
+    auto p_handle = reinterpret_cast<cugraph::c_api::cugraph_resource_handle_t const*>(handle);
+    auto host_srcs =
+      reinterpret_cast<const cugraph::c_api::cugraph_type_erased_host_array_view_t*>(srcs);
+    auto new_device_srcs = new cugraph::c_api::cugraph_type_erased_device_array_t(
```

Comment:
I would be inclined to do this a bit differently.  The values `new_device_srcs`, `new_device_dsts`, `new_device_wgts` and `new_device_counts` that you're creating here are  a potential for a memory leak, since they are not in RAII objects.

You could:
1. Do type dispatching and create these as `rmm::device_uvector<appropriate_type>` and then convert to the return value.  This would mimic the pattern of the algorithm implementation.
2. Create these as `std::unique_ptr` values and release them when you convert to the return value
3. Create the return value as a `std::unique_ptr` (releasing upon return) with the new calls for these operators inside and then just use the values from the return object.  We do know the sizes a priori

Reply:
Instead of this suggestion, I changed the implementation to use other cugraph C API calls (as suggested below), which included using `cugraph_type_erased_device_array_create_from_view()` from #2712 . Let me know if you'd prefer I use the `rmm:device_uvector`/type dispatching approach instead.